contrib/99designs/gqlgen/tracer.go

```diff
@@ -0,0 +1,124 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+// Package gqlgen contains an implementation of a gqlgen tracer, and functions to construct and configure the tracer.
+// The tracer can be passed to the gqlgen handler (see package github.com/99designs/gqlgen/handler)
+//
+// When enabling introspection, please note that introspection queries may cause large traces to be created.
+package gqlgen
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/99designs/gqlgen/graphql"
+	"github.com/vektah/gqlparser/v2/ast"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+const (
+	graphQLQuery = "graphql.query"
+
+	readOp       = "graphql.read"
+	parsingOp    = "graphql.parse"
+	validationOp = "graphql.validate"
+)
+
+type gqlTracer struct {
+	cfg *config
+}
+
+// NewTracer creates a graphql.HandlerExtension instance that can be used with
+// a graphql.handler.Server.
+// Options can be passed in for further configuration.
+func NewTracer(opts ...Option) graphql.HandlerExtension {
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	return &gqlTracer{cfg: cfg}
+}
+
+func (t *gqlTracer) ExtensionName() string {
+	return "DatadogTracing"
+}
+
+func (t *gqlTracer) Validate(schema graphql.ExecutableSchema) error {
+	return nil // unimplemented
+}
+
+func (t *gqlTracer) InterceptResponse(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
+	opts := []ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeGraphQL),
+		tracer.ServiceName(t.cfg.serviceName),
+	}
+	if !math.IsNaN(t.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, t.cfg.analyticsRate))
+	}
+	var (
+		octx *graphql.OperationContext
+	)
+	name := ext.SpanTypeGraphQL
+	if graphql.HasOperationContext(ctx) {
+		octx = graphql.GetOperationContext(ctx)
+		if octx.Operation != nil {
+			if octx.Operation.Operation == ast.Subscription {
+				// These are long running queries for a subscription,
+				// remaining open indefinitely until a subscription ends.
+				// Return early and do not create these spans.
+				return next(ctx)
+			}
+			name = fmt.Sprintf("%s.%s", ext.SpanTypeGraphQL, octx.Operation.Operation)
+		}
+		opts = append(opts, tracer.ResourceName(octx.OperationName))
+		if octx.RawQuery != "" {
+			opts = append(opts, tracer.Tag(graphQLQuery, octx.RawQuery))
+		}
+		for key, val := range octx.Variables {
+			opts = append(opts, tracer.Tag(fmt.Sprintf("graphql.variables.%s", key), val))
+		}
+		opts = append(opts, tracer.StartTime(octx.Stats.OperationStart))
+	}
+	if s, ok := tracer.SpanFromContext(ctx); ok {
+		opts = append(opts, tracer.ChildOf(s.Context()))
```

Comment:
I'm a bit confused here what the intent is. As written this looks for a span in `ctx`, if it finds one then it ignores and adds a `ChildOf` option that uses s.Context() instead?

Then down on 94, `StartSpanFromContext` will end up ignoring the `ChildOf` option since it prefers to use a parent span from `ctx`?

Reply:
Ah this was me just not understanding the API when I originally wrote the code. Forgot to go back and fix that :)